### ISSUE TITLE
Improve #52: separate test executable for c++ and cuda

### ISSUE THREAD

**Issue by valassi** [madgraph5/madgraph4gpu#83]
Hi, sorry for the delay, I only had time to look at #52 now.

I have one problem with these changes: there is a single runTest.exe for cuda and c++. I would very strongly suggest to have two separate executables, just like we are doing with check.exe and gcheck.exe. The two worlds, C++ and CUDA, are not meant to talk to each other. Either we run cuda, or we run c++. Mixing them together is error prone in various ways.

Some of these issues from mixing I had seen in the past (I do not even remember what they were) and they are the reason why I had gone for two separate CPPProcess.o, gCPPProcess.o, and even two namespaces for extra security.

Some mixing issues are more complex to foresee: for instance, I just gave it a try to add OpenMP as discussed in #82, and this now fails linking runTest.exe, because nvcc does not understand OpenMP (and it has no need to understand it).

More fundamentally, if we are testing the C++ implementation, I would build (ie also link) the test with gcc, not with nvcc.

What is happening now is quite confusing, the .o is built with gcc while the .exe is built with nvcc. I would build two separate .o and two separate .exe.
```
/cvmfs/sft.cern.ch/lcg/releases/gcc/8.3.0-cebb0/x86_64-centos7/bin/g++  -O3 -std=c++11 -I. -I../../src -I../../../../../tools/ -I../../../../../tools//googletest/googletest/include/ -DUSE_NVTX -Wall -Wshadow  -I/usr/local/cuda-11.1/include/ -c runTest.cc -o runTest.o

ln -sf runTest.cc runTest_tmp.cu

/usr/local/cuda-11.1/bin/nvcc -o runTest.exe CPPProcess.o runTest.o gCPPProcess.o runTest_tmp.cu  -O3 -std=c++14 -I. -I../../src -I../../../../../tools/ -I../../../../../tools//googletest/googletest/include/ -I/usr/local/cuda-11.1/include/ -DUSE_NVTX -arch=compute_70 -use_fast_math -lineinfo  -ldl -L../../lib -lmodel_sm -L../../../../../tools//googletest/build/lib// -lgtest -lgtest_main -L/usr/local/cuda-11.1/lib64/ -lcuda -lcurand -lcuda

unlink runTest_tmp.cu
```

In summary, I would go for 
- test.o and test.exe built for c++ with gcc
- gtest.o and gtest.exe built for cuda with nvcc

@hageboeck can you implement that? Thanks
Andrea

**Comment by valassi** [comment]
I am no longer so sure of what I was writing yesterday - sorry for the noise/spam.

I realised that in some cases we do need to mix c++ and cuda, so doing this in the tests is acceptable.

In particular, I opened issue #85 and PR #87 for heterogeneous Madgraph doing work both on the CPU (through c++ code) and the GPU (through CUDA code). The final executable must be built with nvcc, not gcc, otherwise some libraries are missing. Since the c++ modules use OMP, the executable links -lgomp, which nvcc understands (while it does not understand -fopenmp).

Similarly, in my OMP branches (issue83 and het) I fixed runTest builds by adding -lgomp. This is enough.

There is nothing else to be done here.. sorry for the noise again.
